distributed/shuffle/tests/test_rechunk.py

```diff
@@ -1275,3 +1271,43 @@ async def test_partial_rechunk_homogeneous_distribution(c, s, *workers):
     nchunks = [len(w.data.keys() & out_keys) for w in workers]
     # There are 40 output chunks and 5 workers. Expect exactly 8 chunks per worker.
     assert nchunks == [8, 8, 8, 8, 8]
+
+
+@gen_cluster(client=True, nthreads=[], config={"optimization.fuse.active": False})
+async def test_partial_rechunk_taskgroups(c, s):
+    arr = da.random.random((40, 40), chunks=(4, 4))
+    arr = arr.rechunk((1, -1), method="p2p")
+
+    _ = c.compute(arr)
+    await async_poll_for(
+        lambda: any(
+            isinstance(task, str) and task.startswith("shuffle-barrier")
+            for task in s.tasks
+        ),
+        timeout=5,
+    )
+    # Ensure that independent partial rechunks do not blow up the number of task groups
+    assert (
+        sum(
+            1
+            for task in s.task_groups
+            if isinstance(task, str) and task.startswith("rechunk-transfer")
+        )
+        == 1
+    )
+    assert (
+        sum(
+            1
+            for task in s.task_groups
+            if isinstance(task, str) and task.startswith("shuffle-barrier")
+        )
+        == 1
+    )
+    assert (
+        sum(
+            1
+            for task in s.task_groups
+            if isinstance(task, str) and task.startswith("rechunk-p2p")
+        )
+        == 1
+    )
```

Comment:
Another perhaps less brittle (but also less powerful) test here would be to assert that there are not too many task groups, something like this maybe (untested)

```suggestion
    )
    
    assert len(s.task_groups) < 10
```

I find that these sorts of tests are nice because they tend to survive in the codebase a bit longer, and be easier to understand by future maintainers (speaking for myself as an example, I understand this test more easily than I understand the tests above.

Reply:
You can also do both.  Doesn't have to be either or.  I just like having loose/generic tests, so if it's easy to add them I think it's good to do so.

(early on in dask development I did the opposite and regretted it when most PRs became 70% changing semi-related tests)

Reply:
I've adjusted the test, the original reproducer creates 27x the number of task groups.